api/repositories/k8s_client.go

```diff
@@ -48,7 +48,7 @@ func (f UnprivilegedClientFactory) BuildClient(authInfo authorization.Info) (cli
 		config.KeyData = pem.EncodeToMemory(keyBlock)
 
 	default:
-		return nil, authorization.NotAuthenticatedError{}
```

Comment:
We're dealing with the `authorization.Info` object here. It has been created from [parsing the `Authorization` header in the middleware](https://github.com/cloudfoundry/cf-k8s-controllers/blob/d97da7815569d4e8f2c4a753bb70160b93c984c0/api/apis/authentication_middleware.go#L45). The logic there is if the user does not supply a `Bearer` or `ClientCert` in the `Authorization` header, then no accepted authentication method has been supplied, hence a `NotAuthenticatedError` is returned. But if they do supply one, but it isn't the correct format, an `InvalidAuthError` is returned. Note that these names match the user errors returned by the API.

The logic here duplicates that in the middleware somewhat. If the `authInfo` is not using either the bearer or the cert scheme, then no authentication has been passed, yet we are here because we need authentication, hence a `NotAuthenticatedError` is appropriate, and matches the API error returned to the user.

If we get past this point, but the supplied token or cert is not valid when checked by k8s, we return an `InvalidAuthError`.

So I don't recommend making the change you are suggesting here.

Reply:
I tend to agree with @kieron-dev's comment above, but that only means that the `authorization.InfoParser` used by the middleware should only ever return `NotAuthenticatedError`, which is [not currently the case](https://github.com/cloudfoundry/cf-k8s-controllers/blob/e05f2ca5c210e539249747fe497d1227a8e47674/api/authorization/info_parser.go#L35)

Reply:
Just to make myself clear I think in this specific instance the error should continue being a `NotAuthenticatedError`